src/ICSharpCode.SharpZipLib/Tar/TarBuffer.cs

```diff
@@ -550,36 +606,77 @@ private void WriteFinalRecord()
 			{
 				int dataBytes = currentBlockIndex * BlockSize;
 				Array.Clear(recordBuffer, dataBytes, RecordSize - dataBytes);
-				WriteRecord();
+				await WriteRecordAsync(ct, isAsync);
 			}
 
-			outputStream.Flush();
+			if (isAsync)
+			{
+				await outputStream.FlushAsync(ct);
+			}
+			else
+			{
+				outputStream.Flush();
+			}
```

Comment:
Why is only this call branched? Shouldn't `await outputStream.FlushAsync(ct);` work fine for both cases? The default implementation of `FlushAsync` just calls `Flush` if not overriden.

Reply:
The call to `WriteRecordAsync` (line 609) isnt branched as the `isAsync` parameter is passed down and it is handled internally.

There are 2 reasons to branch on `Flush` /  `FlushAsync`:
1. `FlushAsync` returns a `Task` and in the case of synchronous access there will be an unnecessary allocation here.
2. (More importantly) If the stream fully supports async usage but the caller explicitly calls the synchronous method it *might* be possible to [end up deadlocking](https://blog.stephencleary.com/2012/07/dont-block-on-async-code.html) so I feel it is probably safer to branch.